fuel-vm/src/interpreter/flow.rs

```diff
@@ -149,138 +161,385 @@ where
+            .copied()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, M))?;
+
+        let b = self
+            .registers
+            .get(rb.to_u8() as usize)
+            .copied()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, M))?;
+
+        let c = self
+            .registers
+            .get(rc.to_u8() as usize)
+            .copied()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, M))?;
 
-        let call = Call::try_from(&self.memory[a as usize..])?;
-        let asset_id =
-            AssetId::try_from(&self.memory[c as usize..cx as usize]).expect("Unreachable! Checked memory range");
+        let d = self
+            .registers
+            .get(rd.to_u8() as usize)
+            .copied()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, M))?;
+
+        self._prepare_call(a, b, c, d)
+    }
+
+    pub(crate) fn call(&mut self, a: Word, b: Word, c: Word, d: Word) -> Result<ProgramState, RuntimeError> {
+        self._prepare_call(a, b, c, d)?;
+        self.run_call()
+    }
+}
+
+#[cfg_attr(test, derive(Default))]
+struct PrepareCallParams {
+    /// A
+    pub call_params_mem_address: Word,
+    /// B
+    pub amount_of_coins_to_forward: Word,
+    /// C
+    pub asset_id_mem_address: Word,
+    /// D
+    pub amount_of_gas_to_forward: Word,
+}
+
+struct PrepareCallReadRegisters<'a> {
+    hp: Reg<'a, HP>,
+    sp: RegMut<'a, SP>,
+    ssp: RegMut<'a, SSP>,
+    fp: RegMut<'a, FP>,
+    pc: RegMut<'a, PC>,
+    is: RegMut<'a, IS>,
+    bal: RegMut<'a, BAL>,
+    cgas: RegMut<'a, CGAS>,
+    ggas: RegMut<'a, GGAS>,
+}
+struct PrepareCallRegisters<'a> {
+    read_registers: PrepareCallReadRegisters<'a>,
+    write_registers: WriteRegistersRef<'a>,
+    unused_registers: PrepareCallUnusedRegisters<'a>,
+}
 
-        let mut frame = self.call_frame(call, asset_id)?;
+struct PrepareCallUnusedRegisters<'a> {
+    zero: Reg<'a, ZERO>,
+    one: Reg<'a, ONE>,
+    of: Reg<'a, OF>,
+    err: Reg<'a, ERR>,
+    ret: Reg<'a, RET>,
+    retl: Reg<'a, RETL>,
+    flag: Reg<'a, FLAG>,
+}
 
-        self.dependent_gas_charge(self.gas_costs.call, frame.total_code_size())?;
+impl<'a> PrepareCallRegisters<'a> {
+    fn copy_registers(&self) -> [Word; VM_REGISTER_COUNT] {
+        copy_registers(&self.into(), &self.write_registers)
+    }
+}
 
-        if self.is_external_context() {
-            self.external_asset_id_balance_sub(&asset_id, b)?;
+struct PrepareCallMemory<'a> {
+    memory: &'a mut [u8; VM_MEMORY_SIZE],
+    call_params: CheckedMemValue<Call>,
+    asset_id: CheckedMemValue<AssetId>,
+}
+
+struct PrepareCallInput<'vm, S> {
+    params: PrepareCallParams,
+    registers: PrepareCallRegisters<'vm>,
+    memory: PrepareCallMemory<'vm>,
+    context: &'vm mut Context,
+    gas_cost: DependentCost,
+    runtime_balances: &'vm mut RuntimeBalances,
+    storage: &'vm mut S,
+    input_contracts: Vec<fuel_types::ContractId>,
```

Comment:
It seems we can replace a vector with an iterator. But we can do that in the follow-up PR

Reply:
Yeh but it's a pain and it's not really that exspensive to copy `ContractId`